Content.Server/GameObjects/Components/Chemistry/MovespeedStatusComponent.cs

```diff
@@ -0,0 +1,61 @@
+using Content.Shared.Movement.Components;
+using Robust.Shared.GameObjects;
+using Robust.Shared.ViewVariables;
+using System;
+using System.Threading;
+
+namespace Content.Server.GameObjects.Components.Chemistry
+{
+    //TODO: refactor movement modifier component because this is a pretty poor solution
+    [RegisterComponent]
+    class MovespeedStatusComponent : Component, IMoveSpeedModifier
+    {
+        
+        public override string Name => "ChemicalMovementSpeedChangeStatus";
+
+        [ViewVariables]
+        private float walkSpeedModifier;
+        public float WalkSpeedModifier
+        {
+            get => walkSpeedModifier;
+            set
+            {
+                walkSpeedModifier = value;
+                ResetTimer();
+            }
+        }
+        [ViewVariables]
+        private float sprintSpeedModifier;
+        public float SprintSpeedModifier
+        {
+            get => sprintSpeedModifier;
+            set
+            {
+                sprintSpeedModifier = value;
+                ResetTimer();
+            }
+        }
+
+        public int EffectTime { get; set; }
+
+        private readonly CancellationTokenSource _cancellation = new CancellationTokenSource();
+
+        public override void Initialize()
+        {
+            base.Initialize();
+            Owner.SpawnTimer(EffectTime, ResetModifiers, _cancellation.Token);
+        }
+
+        private void ResetModifiers()
+        {
+            WalkSpeedModifier = 0;
+            SprintSpeedModifier = 0;
+        }
+
+        private void ResetTimer()
+        {
+            _cancellation.Cancel();
+            Owner.SpawnTimer(EffectTime, ResetModifiers, _cancellation.Token);
```

Comment:
You should add the remaining time onto the timer. You can probably just store when the token was first created and then work out current time less when it was created. In ResetModifiers you just set it to null, then in ResetTimer you can set the field to currenttime if it is null.

Reply:
should I though, because we're gonna be making a new system for this.